magpie/services.py

```diff
@@ -72,85 +65,315 @@ def __init__(self, service, request):
+        resolution (Allow/Deny) provided as output, unless a specific subset of permissions is requested using
+        :paramref:`permissions`. Other permissions are ignored in this case to only resolve requested ones.
+        For example, this parameter can be used to request only ACL resolution from specific permissions applicable
+        for a given request, as obtained by :meth:`permission_requested`.
 
-    def expand_acl(self, resource, user):
-        # type: (models.Resource, models.User) -> None
-        if resource:
-            for ace in resource.__acl__:
-                self.acl.append(ace)
+        Permissions scoped as `match` can be ignored using :paramref:`allow_match`, such as when the targeted resource
+        does not exist.
 
-            if user:
-                permissions = ResourceService.perms_for_user(resource, user, db_session=self.request.db)
-                for outcome, perm_user, perm_name in permission_to_pyramid_acls(permissions):
-                    self.acl.append((outcome, perm_user, perm_name,))
+        .. seealso::
+            - :meth:`ServiceInterface.resource_requested`
+        """
+        if not permissions:
+            permissions = self.allowed_permissions(resource)
+        requested_perms = set(permissions)  # type: Set[Permission]
+        effective_perms = dict()            # type: Dict[Permission, PermissionSet]
+
+        # immediately return all permissions if user is an admin
+        db_session = self.request.db
+        admin_group = get_constant("MAGPIE_ADMIN_GROUP", self.request)
+        admin_group = GroupService.by_group_name(admin_group, db_session=db_session)
+        if admin_group in user.groups:  # noqa
+            return [PermissionSet(perm, Access.ALLOW, Scope.MATCH, PermissionType.EFFECTIVE) for perm in permissions]
+
+        # current and parent resource(s) recursive-scope
+        match = allow_match
+        while resource is not None:  # bottom-up until service is reached
+
+            # include both permissions set in database as well as defined directly on resource
+            cur_res_perms = ResourceService.perms_for_user(resource, user, db_session=db_session)
+            cur_res_perms.extend(permission_to_pyramid_acls(resource.__acl__))
+
+            for perm_name in requested_perms:
+                for perm_tup in cur_res_perms:
+                    perm_set = PermissionSet(perm_tup)
+
+                    # if user is owner (directly or via groups), all permissions are set,
+                    # but continue processing this resource until end in case user explicit deny reverts it
+                    if perm_tup.perm_name == ALL_PERMISSIONS:
+                        for perm in requested_perms:
+                            all_perm = PermissionSet(perm, perm_set.access, perm.scope, perm.type)
+                            if perm_set.access == Access.DENY:
```

Comment:
I don't want to be picky, but this function is the most interesting part of this PR.
I would like to have a little bit more comment on this if/else.
The if is about a user that own a ressource but is denied access to it?

Reply:
That would be the idea, but it is not really doing anything at the moment since there is no actual owner currently. 
I can add a FIXME note for when/if ownership rules get implemented properly.